File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
  def numSquares(self, n: int) -> int:
    MAX = int(1e5+1)
    squares = [i*i for i in range(1, int(n**0.5)+1)]
    dp = [float('inf')] * (n + 1)
    for square in squares:
      dp[square] = 1
      for i in range(square+1, n+1):
        dp[i] = min(dp[i], dp[i-square] + 1)
    return dp[n]
    

File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {
public:
  int numSquares(int n) {
    vector<int> dp(n+1,10000000);
    dp[0] = 0;
    for (int i = 1; i <= n; ++i) {
      for (int j = 1; j*j <= i; ++j) {
        dp[i] = min(dp[i], 1+dp[i-j*j]);
      }
    }
    return dp[n];
  }
};